crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+    pub(crate) fn first_party(path: &'a (impl AsRef<FileSystemPath> + ?Sized)) -> Option<Self> {
+        let path = path.as_ref();
+        if path
+            .extension()
+            .map_or(true, |ext| matches!(ext, "pyi" | "py"))
+        {
+            Some(Self::first_party_unchecked(path))
+        } else {
+            None
+        }
+    }
+
+    #[must_use]
+    #[allow(unsafe_code)]
+    fn first_party_unchecked(path: &'a (impl AsRef<FileSystemPath> + ?Sized)) -> Self {
+        // SAFETY: FirstPartyPath is marked as #[repr(transparent)] so the conversion from a
+        // *const FileSystemPath to a *const FirstPartyPath is valid.
+        Self::FirstParty(unsafe {
+            &*(path.as_ref() as *const FileSystemPath as *const FirstPartyPath)
+        })
+    }
+
+    #[must_use]
+    pub(crate) fn standard_library(
+        path: &'a (impl AsRef<FileSystemPath> + ?Sized),
+    ) -> Option<Self> {
+        let path = path.as_ref();
+        // Unlike other variants, only `.pyi` extensions are permitted
+        if path.extension().map_or(true, |ext| ext == "pyi") {
+            Some(Self::standard_library_unchecked(path))
+        } else {
+            None
+        }
+    }
+
+    #[must_use]
+    #[allow(unsafe_code)]
+    fn standard_library_unchecked(path: &'a (impl AsRef<FileSystemPath> + ?Sized)) -> Self {
+        // SAFETY: StandardLibraryPath is marked as #[repr(transparent)] so the conversion from a
+        // *const FileSystemPath to a *const StandardLibraryPath is valid.
+        Self::StandardLibrary(unsafe {
+            &*(path.as_ref() as *const FileSystemPath as *const StandardLibraryPath)
+        })
+    }
+
+    #[must_use]
+    pub(crate) fn site_packages(path: &'a (impl AsRef<FileSystemPath> + ?Sized)) -> Option<Self> {
+        let path = path.as_ref();
+        if path
+            .extension()
+            .map_or(true, |ext| matches!(ext, "pyi" | "py"))
+        {
+            Some(Self::site_packages_unchecked(path))
+        } else {
+            None
+        }
+    }
+
+    #[must_use]
+    #[allow(unsafe_code)]
+    fn site_packages_unchecked(path: &'a (impl AsRef<FileSystemPath> + ?Sized)) -> Self {
+        // SAFETY: SitePackagesPath is marked as #[repr(transparent)] so the conversion from a
+        // *const FileSystemPath to a *const SitePackagesPath is valid.
+        Self::SitePackages(unsafe {
+            &*(path.as_ref() as *const FileSystemPath as *const SitePackagesPath)
+        })
+    }
+
+    #[must_use]
+    fn load_typeshed_versions(db: &dyn Db, stdlib_root: &StandardLibraryPath) -> TypeshedVersions {
+        let StandardLibraryPath(stdlib_fs_path) = stdlib_root;
+        let versions_path = stdlib_fs_path.join("VERSIONS");
+        let versions_file = system_path_to_file(db.upcast(), versions_path).unwrap();
+        parse_typeshed_versions(db, versions_file)
+    }
+
+    #[must_use]
+    pub(crate) fn is_directory(&self, db: &dyn Db, search_path: impl Into<Self>) -> bool {
```

Comment:
I'm leaning towards removing all methods that take a `&db` out of path and instead implement them as methods on `ModuleResolver`. It just feels a bit heavyweight to me that an `is_directory` call ends up loading the type shed versions. It also makes it harder to optimize caching of typeshed versions; e.g. the module resolver could load the typeshed versions once and cache them for the entire resolution of a `resolve_module` call where path needs to call `parse_typeshed_versions` for every `is_directory`, `is_regular_package`, ...  call

Reply:
I see your point. However, the advantage of doing this as late as possible is that we only parse the typeshed versions when we actually need to: some module resolutions won't require us to look at typeshed `VERSIONS` at all, whereas if the method requires `TypeshedVersions` as an argument then we need to parse the `VERSIONS` file eagerly whether or not it ends up being resolved using a standard-library search path. `parse_typeshed_versions` is a Salsa-tracked function, as well, so the results should be memoized after the first call until the `VERSIONS` file is edited or deleted by the user from the custom typeshed directory (which should be very rare).

Reply:
> I agree, we should avoid resolving `TypeshedVersions` if they aren't needed. I think you can use an `Option` or `OnceCell` to cache the versions lazily. [doc.rust-lang.org/std/cell/struct.OnceCell.html](https://doc.rust-lang.org/std/cell/struct.OnceCell.html)

Ahh, that's very clever. Thanks!

> I agree, we should avoid resolving `TypeshedVersions` if they aren't needed. I think you can use an `Option` or `OnceCell` to cache the versions lazily. [doc.rust-lang.org/std/cell/struct.OnceCell.html](https://doc.rust-lang.org/std/cell/struct.OnceCell.html)
> 
> > parse_typeshed_versions is a Salsa-tracked function,
> 
> That's correct, but caching still has a non-trivial overhead of doing a lookup in a concurrent hash map and then checking if the revision is still fresh.

Hmm, fair enough. I guess I'm used to Python, where looking things up in a cache is ~free for a method decorated with `@functools.lru_cache` :-) But I guess in Rust, ~everything's faster than Python, so the cost of looking things up in a cache becomes significant.

Reply:
It took me a while to figure out how to do this, as the lifetimes got quite complicated. In the end I succeeded, but I had to reintroduce an `Arc` in the `TypeshedVersions` struct rather than have the Salsa query return a reference.